# libsmb2 server, directory listings (MS-FSCC 2.4): NextEntryOffset is relative to the entry, and every
# entry starts on an 8-byte boundary. libsmb2 wrote the absolute offset and padded entries to 4 bytes; its own
# client copes with the padding, Windows does not ("The specified server cannot perform the requested
# operation", ERROR_BAD_NET_RESP).
--- a/lib/smb2-cmd-query-directory.c
+++ b/lib/smb2-cmd-query-directory.c
@@ -249,10 +249,10 @@
                                 switch (info_class)
                                 {
                                 case SMB2_FILE_ID_FULL_DIRECTORY_INFORMATION:
-                                        fs_size = PAD_TO_32BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                        fs_size = PAD_TO_64BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
                                         break;
                                 case SMB2_FILE_ID_BOTH_DIRECTORY_INFORMATION:
-                                        fs_size = PAD_TO_32BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                        fs_size = PAD_TO_64BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
                                         break;
                                 default:
                                         fs_size = 0;
@@ -316,10 +316,10 @@
                         switch (info_class)
                         {
                         case SMB2_FILE_ID_FULL_DIRECTORY_INFORMATION:
-                                fs_size = PAD_TO_32BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                fs_size = PAD_TO_64BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
                                 break;
                         case SMB2_FILE_ID_BOTH_DIRECTORY_INFORMATION:
-                                fs_size = PAD_TO_32BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                fs_size = PAD_TO_64BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
                                 break;
                         default:
                                 fs_size = 0;
@@ -328,7 +328,7 @@
                         in_offset += PAD_TO_64BIT(sizeof(struct smb2_fileidbothdirectoryinformation));
                         in_remain -= PAD_TO_64BIT(sizeof(struct smb2_fileidbothdirectoryinformation));
                         if (in_remain >= SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE) {
-                                smb2_set_uint32(iov, offset + 0, offset + fs_size);
+                                smb2_set_uint32(iov, offset + 0, fs_size);
                         }
                         else {
                                 smb2_set_uint32(iov, offset + 0, 0);
@@ -337,7 +337,7 @@
                         switch (info_class)
                         {
                         case SMB2_FILE_ID_FULL_DIRECTORY_INFORMATION:
-                                fs_size = PAD_TO_32BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                fs_size = PAD_TO_64BIT(SMB2_FILEID_FULL_DIRECTORY_INFORMATION_SIZE + fname_len);
                                 smb2_set_uint32(iov, offset + 4, fs->file_index);
                                 smb2_set_uint64(iov, offset + 8, smb2_timeval_to_win(&fs->creation_time));
                                 smb2_set_uint64(iov, offset + 16, smb2_timeval_to_win(&fs->last_access_time));
@@ -355,7 +355,7 @@
                                 }
                                 break;
                         case SMB2_FILE_ID_BOTH_DIRECTORY_INFORMATION:
-                                fs_size = PAD_TO_32BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
+                                fs_size = PAD_TO_64BIT(SMB2_FILEID_BOTH_DIRECTORY_INFORMATION_SIZE + fname_len);
                                 smb2_set_uint32(iov, offset + 4, fs->file_index);
                                 smb2_set_uint64(iov, offset + 8, smb2_timeval_to_win(&fs->creation_time));
                                 smb2_set_uint64(iov, offset + 16, smb2_timeval_to_win(&fs->last_access_time));
